llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -3024,3 +3024,170 @@ mccasformats::v1::loadDIETopLevel(DIETopLevelRef TopLevelRef) {
+  return AbbrevEntryReader(AbbrevData);
+}
+
+Error DIEVisitor::visitDIERef(BinaryStreamReader &DataReader,
+                              unsigned AbbrevIdx, StringRef DIEData,
+                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
+  AbbrevEntryReader AbbrevReader =
+      getAbbrevEntryReader(AbbrevEntries, AbbrevIdx);
+
+  if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
+    StartTagCallback(*MaybeTag, AbbrevIdx);
+  else
+    return MaybeTag.takeError();
+
+  Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
+  if (!MaybeHasChildren)
+    return MaybeHasChildren.takeError();
+
+  if (auto E = visitDIEAttrs(AbbrevReader, DataReader, DIEData))
+    return E;
+
+  if (!*MaybeHasChildren) {
+    EndTagCallback(false /*HadChildren*/);
+    return Error::success();
+  }
+
+  while (true) {
+    Expected<uint64_t> ChildAbbrevIdx = readAbbrevIdx(DistinctReader);
+    if (!ChildAbbrevIdx)
+      return ChildAbbrevIdx.takeError();
+
+    if (*ChildAbbrevIdx == getEndOfDIESiblingsMarker())
+      break;
+
+    if (*ChildAbbrevIdx == getDIEInAnotherBlockMarker()) {
+      if (auto E = visitDIERef(DIEChildrenStack.front()))
+        return E;
+      DIEChildrenStack = DIEChildrenStack.drop_front();
+      continue;
+    }
+
+    if (auto E =
+            visitDIERef(DataReader, *ChildAbbrevIdx, DIEData, DIEChildrenStack))
+      return E;
+  }
+
+  EndTagCallback(true /*HadChildren*/);
+  return Error::success();
+}
+
+Error DIEVisitor::visitDIERef(DIEDataRef Ref) {
+  StringRef DIEData = Ref.getData();
+  BinaryStreamReader DataReader(DIEData, support::endianness::little);
+
+  Expected<uint64_t> MaybeAbbrevIdx = readAbbrevIdx(DistinctReader);
+  if (!MaybeAbbrevIdx)
+    return MaybeAbbrevIdx.takeError();
+  auto AbbrevIdx = *MaybeAbbrevIdx;
+
+  // The tag of a fresh block must be meaningful, otherwise we wouldn't have
+  // made a new block.
+  assert(AbbrevIdx != getEndOfDIESiblingsMarker() &&
+         AbbrevIdx != getDIEInAnotherBlockMarker());
+
+  NewBlockCallback(Ref.getID().toString());
+
+  Expected<SmallVector<DIEDataRef>> MaybeChildren =
+      loadAllRefs<DIEDataRef>(Ref);
+  if (!MaybeChildren)
+    return MaybeChildren.takeError();
+  ArrayRef<DIEDataRef> Children = *MaybeChildren;
+
+  return visitDIERef(DataReader, AbbrevIdx, DIEData, Children);
+}
+
+Error mccasformats::v1::visitDebugInfo(
+    Expected<DIETopLevelRef> MaybeTopLevelRef,
+    std::function<void(StringRef)> HeaderCallback,
+    std::function<void(dwarf::Tag, uint64_t)> StartTagCallback,
+    std::function<void(dwarf::Attribute, dwarf::Form, StringRef, bool)>
+        AttrCallback,
+    std::function<void(bool)> EndTagCallback,
+    std::function<void(StringRef)> NewBlockCallback) {
+
+  Expected<LoadedDIETopLevel> LoadedTopRef =
+      loadDIETopLevel(std::move(MaybeTopLevelRef));
+  if (!LoadedTopRef)
+    return LoadedTopRef.takeError();
+
+  BinaryStreamReader DistinctReader(LoadedTopRef->DistinctData.getData(),
```

Comment:
Should we assume the endianness to be little always? Isn't it stored in the MCObjectProxy?

`  if (HeaderRef::Cast(MCObj)) {
    MachOHeaderParser P;
    if (Error Err = P.parse(MCObj.getData()))
      return Err;
    Is64Bit = P.Is64Bit;
    IsLittleEndian = P.IsLittleEndian;`